book/continuous_integration.md

```diff
@@ -0,0 +1,192 @@
+
+Two main types of tool exist for these tasks:
+
+* Linters - these analyse your code to flag stylistic errors (and sometimes bugs or security issues too).
+* Formatters - these not only detect when you have diverged from a style, but will automatically correct the formatting of your code to conform to a particular style.
+
+```{list-table} Packages that can be used for linting or formatting in Python and R
+:header-rows: 1
+:name: linters
+
+* - Language
+  - Linters
+  - Formatters
+* - Python
+  - `flake8`, `pylint`, `Bandit`
+  - `Black`, `Isort`
+* - R
+  - `lintr`
+  - `formatR`, `styler`
+```
+
+Be sure to read the documentation for any of these tools, to understand what they are checking or changing in your code. Some can be configured to ignore or detect specific types of formatting error. You can run multiple of these, to catch a broader range of stylistic or programmatic errors.
+
+## Example use cases for GitHub Actions
+
+The following examples are presented to support understanding of the topic discussed above.
+
+### Configure GitHub actions to automate tests
+
+Below is an example configuration file, for use with GitHub actions. The `YAML` file format, used below, is common to a number of other CI tools.
+
+```yaml
+name: Test python package
+
+on:
+  push:
+    branches:
+      - master
+  pull_request:
+
+jobs:
+  build:
+
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        python-version: [3.6, 3.7, 3.8, 3.9]
+
+    steps:
+    - uses: actions/checkout@v2
+
+    - name: Set up Python ${{ matrix.python-version }}
+      uses: actions/setup-python@v2
+      with:
+        python-version: ${{ matrix.python-version }}
+
+    - name: Install dependencies
+      run: |
+        python -m pip install --upgrade pip
+        pip install pytest 
+        pip install -r requirements.txt
+        
+    - name: Test with pytest
+      run: |
+        pytest
+```
+
+The first section of this example describes when our workflow should be run. In this case, we're running the CI workflow whenever code is `push`ed to the `master` branch or where any Pull Request is created. In the case of Pull Requests, the results of the CI workflow will be report on the request's page. If any of the workflow stages fail, this can block the merge of these changes onto a more stable branch. Subsequent commits to the source branch will trigger the CI workflow to run again.
+
+Below `jobs`, we're defining what tasks we would like to run when our workflow is triggered. We define what operating system we would like to run our workflow on - the Linux operating system `ubuntu` here. The `matrix` section under `strategy` defines parameters for the workflow. The workflow will be repeated for each combination of parameters supplied here - in this case the 4 latest Python versions.
```

Comment:
These aren't the 4 latest versions of python. 

'in this case 4 recent Python versions.'